### DISCUSSION TITLE
unattended / silent install - it's a lie, Steven!

### DISCUSSION THREAD

**Discussion by working-name** [Homebrew/discussions#3199] [Getting started] [answered]
Hello there,

So... turns out unattended is not unattended. Maybe you should not check if the script is run as `root` when ~~`UNATTENDED=1`~~ `NONINTERACTIVE=1` is passed. As it stands right now unattended option is a lie. If you pass the flag, you don't press ENTER to confirm but you are asked for the root password. If you run the installer as root, it tells you not to 🤦 

**Before the weird "don't do this" police comes along:** yes, I know what I'm doing, it's an Automator flow and the user is well aware it will install brew, and yes, they do want brew installed regularly, not the crippled in a special path type way. The only way to run elevated shell scripts is via Apple Script in an Automator workflow, and Apple Script does not conveniently bring up a terminal window to bug the user. It is already running the shell script as root, after asking for the admin credentials in the native macOS way.

So the question remains: how do you install brew unattended but for real this time? Thanks 😄

**Comment by Bo98** [comment]
> UNATTENDED=1

We don't check this env or even document it anywhere as far as I know? We do check `NONINTERACTIVE` however.

**Reply by working-name** [reply]
Ah, right, tomato potato... `NONINTERACTIVE=1` is what I meant..

**Reply by carlocab** [reply]
Setting `NONINTERACTIVE` doesn't remove the need to authenticate for `sudo`.

You need to provide a way to authenticate non-interactively (e.g. `SUDO_ASKPASS`) or avoid the need for authentication (e.g. editing `/etc/sudoers` -- probably not what I would do here).

**Reply by working-name** [reply]
Unfortunately setting an env var is not a feasible option. Really not the greatest thing to prompt the user "Hey, what's your computer login password again?" ... not exactly confidence inspiring. But thank you, that's a good suggestion.

I would also rather avoid making changes to sudoers but I guess I could make it and revert it. It's a last resort type option.

**Reply by carlocab** [reply]
If you're able to set `NONINTERACTIVE`, why isn't setting `SUDO_ASKPASS` feasible?

**Reply by working-name** [reply]
![image](https://user-images.githubusercontent.com/16505052/164371332-ec889b0d-0580-47ef-b65f-649852da45e1.png)

**Reply by carlocab** [reply]
That's not how `SUDO_ASKPASS` works, I'm afraid. See https://stackoverflow.com/questions/12608293/how-to-setup-a-sudo-askpass-environment-variable.

As an aside:

Please [don't use screenshots](https://meta.stackoverflow.com/a/285557) when sharing the output of commands in your terminal. They're often difficult to read and impossible to reuse (e.g. when looking up error messages). Instead, copy the text here and wrap them in three back-ticks (```) on the line before and on the line after to format them properly.

**Reply by working-name** [reply]
Right, well I still think the same as I did initially: I don't want to ask the user for their login password to place it in a file to then have it be picked up, that's too tacky. The workflow is already asking the user for their login password (native macOS prompt, not scripted by me). Not a solution, just another option which is a no go for me.

**Reply by carlocab** [reply]
> I don't want to ask the user for their login password to place it in a file to then have it be picked up, that's too tacky.

That's not the only way to use `SUDO_ASKPASS`, but ok.

If you don't like the solutions you've been offered, you're welcome to develop another one. If you're feeling generous, you might even consider contributing what you've made.

**Reply by working-name** [reply]
I respectfully disagree that your 2 suggestions qualify as a solution. Furthermore, here's a quote from the SO you linked:

> Basically the value of $SUDO_ASKPASS is to be an executable that will spit your password to standard out when invoked.
> `SUDO_ASKPASS=${HOME}/bin/pw.sh sudo shutdown -h now`

That clearly involves asking the user for their password through some third-party input method, which, repeating myself for the 3rd time, I wish to avoid. You seem to be selectively reading that part for some reason. Maybe you think highly of the askpass method, who knows? Point is, not a solution for me. Thank you, once again, but no.

Will skip over the snarky make your own homebrew if you don't like it. One thing is for sure, that type of attitude will drive donations up.

**Reply by SMillerDev** [reply]
> That clearly involves asking the user for their password through some third-party input method, which, repeating myself for the 3rd time, I wish to avoid. 

Only if you aren't the user and don't have a way to know the password in advance. In that case users might not appropriate you running something as root on their machines though.

> Point is, not a solution for me. Thank you, once again, but no.

That's fine, homebrew does not have to have a solution to every problem. 

> Will skip over the snarky make your own homebrew if you don't like it.

Nobody said anything about making your own homebrew. There was only a suggestion for adding your own solution to the installer. Which is literally how open source is supposed to work.

**Reply by carlocab** [reply]
The SO post I linked to suggests a way to use AppleScript to provide the input for `SUDO_ASKPASS`, which sounds to me to be not all too different from the way you were trying to get users to authenticate in the first place.

If any authentication at all is a dealbreaker for you, then there likely isn't a way to do what you're looking to do.

> Will skip over the snarky make your own homebrew if you don't like it.

That's just the way open source works. Just because we've made our work on Homebrew freely available does not mean that we're also going to work on optimising it for your personal use cases. Especially not after accusations of lying, selective reading, and snark.

I will note, however, that I certainly did not suggest that you make your own Homebrew. Only a very tortured reading of what I actually said would allow one to come to that conclusion.

> One thing is for sure, that type of attitude will drive donations up.

If donations go down because people don't find Homebrew useful enough to make it worth their donations, then I am 100% ok with that. The donations aren't for us: they're there to fund the things we do to make Homebrew better for other users. If other users don't find our work useful, then there is no need for the donations either, because we wouldn't need to do the things we spend the donations on.

**Reply by working-name** [reply]
> which sounds to me to be not all too different from the way you were trying to get users to authenticate in the first place

Sorry, not sure how to explain this any differently, but the assumption is incorrect.

> Only a very tortured reading of what I actually said would allow one to come to that conclusion.

This type of situation (unattended installs) was discussed across multiple tickets, and I concluded that it is against what is considered acceptable to this development team. Hence the only option left for your snarky remark is to make my own tool, since this project does not take PRs to that effect( see https://github.com/Homebrew/brew/issues/3428#issuecomment-343475453 `This is definitely not going to get merged.` )

> If donations go down because people don't find Homebrew useful enough to make it worth their donations, then I am 100% ok with that.

The project's usefulness is not in question here, only your attitude as it relates to and directly affects people's willingness to donate.

---- 

I have noticed that both you and others marked as maintainer, contributor have the same general tepid involvement in people's requests. And that's fine, it is your prerogative. What bugs me is that you insist that my lack of understanding is the issue when in fact it's just an ignorant understanding of native vs applescript prompts and the implications of asking for the same thing twice, which is not only unnecessary, but also worrisome. I get it, you offered options, I explained why I think they're not solutions, but you seem to be highly bothered by my explanations and my refusal to use either one.

**Reply by working-name** [reply]
> Nobody said anything about making your own homebrew. There was only a suggestion for adding your own solution to the installer. Which is literally how open source is supposed to work.

Maybe have a look at the comment I linked above for @carlocab - that does not seem like a welcoming open source project to me.

**Reply by carlocab** [reply]
> Sorry, not sure how to explain this any differently, but the assumption is incorrect.

Ok, that's fine. It looks to me as if the install script isn't for you. You can try using one of the [alternative install methods](https://docs.brew.sh/Installation#alternative-installs). And no, using an alternative install method does not necessarily entail this:

> the crippled in a special path type way

> This type of situation (unattended installs) was discussed across multiple tickets, and I concluded that it is against what is considered acceptable to this development team. Hence the only option left for your snarky remark is to make my own tool, since this project does not take PRs to that effect( see [Homebrew/brew#3428 (comment)](https://github.com/Homebrew/brew/issues/3428#issuecomment-343475453) `This is definitely not going to get merged.` )

Our entire discussion here is about the unattended installs of Homebrew itself. The issue you linked to is about the unattended _use_ of Homebrew, which is a related but a very clearly distinct thing. Taking my comment made in the former context and jamming it into the latter is the very definition of a tortured reading.

> The project's usefulness is not in question here, only your attitude as it relates to and directly affects people's willingness to donate.

This doesn't really change much about what I said. The donations aren't really for us. If users don't want to contribute to making Homebrew better for them, fine with me.

While we're on the subject of attitude: yours relates to and directly affects the time, energy, and motivation of maintainers to keep working on Homebrew. Homebrew will live without donations from users. The same isn't true if maintainers decide to stop working on it, and the attitude you've displayed here isn't exactly doing wonders for maintainer motivation at the moment.

> I have noticed that both you and others marked as maintainer, contributor have the same general tepid involvement in people's requests. And that's fine, it is your prerogative. What bugs me is that you insist that my lack of understanding is the issue when in fact it's just an ignorant understanding of native vs applescript prompts and the implications of asking for the same thing twice, which is not only unnecessary, but also worrisome. I get it, you offered options, I explained why I think they're not solutions, but you seem to be highly bothered by my explanations and my refusal to use either one.

I'm not sure where I insisted your lack of understanding is the issue. If you've taken my comments to mean that, then that's on you.

I'm not bothered by how you decide to use the information available here. What I am bothered about is the accusations you've made here, when at least some of them could very well be levelled against you. I'm not a fan of hypocrisy.

You complain that I've not read what you've said carefully ("selective reading"), and yet you twist my words so that they mean something else entirely.

You complain about snark, and yet the commentary in the title of this discussion alone ("it's a lie, Steven!") easily conveys a healthy amount of snark.

**Reply by working-name** [reply]
> You can try using one of the [alternative install methods](https://docs.brew.sh/Installation#alternative-installs).

Yeah that was what I was reading up on, seems like the only option I have that does not bug the user again.

> unattended _use_ of Homebrew, which is a related but a very clearly distinct thing

The difference is between `install` and `upgrade`, otherwise it's pretty much the same: password needs to be given by the user, therefore still about unattended installs. So I would venture to say less of a tortured reading and more of a practical assessment vs grasping at straws for exoneration.

> While we're on the subject of attitude: yours relates to and directly affects the time, energy, and motivation of maintainers to keep working on Homebrew. Homebrew will live without donations from users. The same isn't true if maintainers decide to stop working on it, and the attitude you've displayed here isn't exactly doing wonders for maintainer motivation at the moment.

And now you finally get it - t's a two-way street. Given that my initial interaction with your project has been a negative one overall, you still have to wonder why the tone? I google up some issues and find contributors being rude and dismissive towards people who even offer financial incentives to get something developed. How much more ridiculous can it get?

You are right though - if homebrew is about alienating people who choose to interact with it, it will come to a point where any developer will find little incentive to continue working on it.

> I'm not sure where I insisted your lack of understanding is the issue. If you've taken my comments to mean that, then that's on you.
> 
> I'm not bothered by how you decide to use the information available here. What I am bothered about is the accusations you've made here, when at least some of them could very well be levelled against you. I'm not a fan of hypocrisy.

Don't misattribute my reaction to homebrew's poor public interactions and misleading documentation to an innate evil action towards the project. That would be incorrect.

> You complain that I've not read what you've said carefully ("selective reading"), and yet you twist my words so that they mean something else entirely.

There's nothing to twist. You provided 2 options, I voiced concern for the first, considered the second as a possibility. Then the discussion went on where I repeated my explanation for why I would avoid 1, leading to your last comment `sounds to me to be not all too different from the way you were trying to get users to authenticate in the first place.` I don't think I was off base in stating that you selectively read what I said, because throughout the discussion the main point I made was ignored: asking the user for their login with a home-baked prompt after the OS itself asks them for it is not a good solution for me.

Of course there is a possibility that you're not a mac user or not quite familiar with the difference between the OS prompt and an applescript one, but I still find it hard to believe that my 3+ attempts at describing said difference were that atrocious.

> You complain about snark, and yet the commentary in the title of this discussion alone ("it's a lie, Steven!") easily conveys a healthy amount of snark.

This is hardly a valid point, but it is not snark I portray but outright indignation. 

It's a lie, Steven is a paraphrased line from a movie (nacho libre), actually a funny line. Either way, it does accurately describe homebrew's documentation which misleads users by stating that `the Homebrew installer [...] doesn’t prompt for passwords` with `NONINTERACTIVE=1`. That should read `skip confirmation to continue installation after listing default paths` at best, or more explicitly say `it does not prevent password prompts for sudo` in which case it isn't unattended by definition.

**Reply by SMillerDev** [reply]
> The difference is between install and upgrade, otherwise it's pretty much the same: password needs to be given by the user, therefore still about unattended installs.

Except it's not. Homebrew only requires `sudo` for the installer. Please stop pretending to know the use cases for homebrew than people who voluntarily spend hours every day to improve the project because you clearly don't. I'm wondering if you ever used homebrew at all with comments like these.

> asking the user for their login with a home-baked prompt after the OS itself asks them for it is not a good solution for me.

Which a sudoers edit was suggested for.

> It's a lie, Steven is a paraphrased line from a movie (nacho libre), actually a funny line.

Ah, because someone said something in a movie it's not snark.

> it does accurately describe homebrew's documentation which misleads users

Or... You feel mislead because it wasn't the solution you were hoping for. You're the first person to suggest it's misleading in the 6 years I've helped with Homebrew. But the beauty of open source, that you seem to be misunderstanding, is that nobody owes you any support but yourself. If you think that line is bad, fix it. Don't waste people's time arguing how they meant their comments (because you're so clearly wrong here).

**Reply by carlocab** [reply]
It's really hilarious how you claim you haven't twisted my words and then proceed to do exactly that at least thrice in the same comment shortly thereafter. I'd offer an explanation for this, but I suspect doing so simply invites more misrepresentation. Fool me twice and all that.

**Reply by working-name** [reply]
Got it, so you're both out of steam, pulling cards like "I know better" and "nobody owes you anything" which are both inane insinuations and nowhere near reality.

**Comment by working-name** [comment]
Anyway I can `perl -pie` your bash script before it runs, but since I'm not familiar with the shell script: what happens if I remove the root check? How does it know for which user to set it up if I run it as root from the beginning?

**Comment by working-name** [comment, marked as answer]
After reviewing https://github.com/Homebrew/brew/issues/3428 I can see that this behavior is marked with misunderstandings and shortsightedness on the part of development. 

~~The least you could do is update the documentation to stop lying to people:~~ Documentation was amended: https://github.com/working-name/brew/commit/12363f673d96c52c503fdc092a8f3a1bbe4c4d8e

>If you want to run the Homebrew installer non-interactively without prompting for passwords (e.g. in automation scripts), you can use:
